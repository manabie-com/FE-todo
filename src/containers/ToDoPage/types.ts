import { EntityState } from '@reduxjs/toolkit';
import { Todo, TodoStatus } from 'models/todo';

export interface ToDoPageState extends EntityState<Todo> {
  filter: string;
}

export type ContainerState = ToDoPageState;

export type EnhanceTodoStatus = TodoStatus | 'ALL';
