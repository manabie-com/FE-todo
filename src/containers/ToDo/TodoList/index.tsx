import React, { useRef, useState } from 'react';

/** Components */
import { Input, Button } from '@components/index';

/** Models */
import { Todo, TodoStatus } from '@models/todo';

/** Utils */
import { isTodoCompleted } from '@utils/index';

/** Actions */
import { deleteTodo, setTodos } from '@store/actions';

/** Hooks */
import useOnClickOutside from '@hooks/useOnClickOutSide';

import Styles from './todoList.module.css';

type ITodoList = {
  onUpdateTodoStatus: any;
  showing: string;
  todos: Todo[] | any;
  dispatch: any;
};

const TodoList = ({
  showing,
  onUpdateTodoStatus,
  todos,
  dispatch,
}: ITodoList) => {
  const refItem = useRef(null);
  const [content, setContent] = useState('');

  const onDeleteTodo = (id: string) => {
    const newTodo = showTodos.filter((todo: Todo) => {
      return todo.id !== id;
    });
    dispatch(deleteTodo(newTodo));
  };

  let showTodos = todos.filter((todo: Todo) => {
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });

  const onEditTodo = (id: string) => {
    setContent(todos.filter((todo: Todo) => todo.id === id)[0].content);
    const newData = todos.map((todo: Todo) => ({
      ...todo,
      isEdit: todo.id === id ? true : false,
    }));
    dispatch(setTodos(newData));
  };

  const onEnterTodo = (e: any, id: string) => {
    if (
      e.key === 'Enter' &&
      todos.filter((todo: Todo) => !Boolean(todo.content)).length === 0
    ) {
      const newData = todos.map((todo: Todo) => ({
        ...todo,
        content: todo.id === id ? content : todo.content,
        isEdit: false,
      }));
      dispatch(setTodos(newData));
      setContent(newData);
    }
  };

  const onChangeItemTodo = (e: any, id: string) => {
    const newData = todos.map((todo: Todo) => ({
      ...todo,
      content: todo.id === id ? e.target.value : todo.content,
    }));
    dispatch(setTodos(newData));
    setContent(e.target.value);
  };

  useOnClickOutside(refItem, () => {
    const newData = todos.map((todo: Todo) => ({
      ...todo,
      isEdit: false,
    }));
    dispatch(setTodos(newData));
  });

  return (
    <div className={Styles.container}>
      {showTodos.length !== 0 ? (
        showTodos.map((todo: Todo, index: number) => {
          return (
            <div className={Styles.todo} key={index}>
              <div className={Styles.nameTodo}>
                <Input
                  className={Styles.checkbox}
                  type='checkbox'
                  checked={isTodoCompleted(todo)}
                  onChange={(e: any) => onUpdateTodoStatus(e, todo.id)}
                />
                {todo.isEdit ? (
                  <Input
                    ref={refItem}
                    value={content}
                    onKeyDown={(e: any) => onEnterTodo(e, todo.id)}
                    onChange={(e: any) => onChangeItemTodo(e, todo.id)}
                  />
                ) : (
                  <span
                    onDoubleClick={
                      todo.status === TodoStatus.ACTIVE
                        ? () => onEditTodo(todo.id)
                        : undefined
                    }
                    className={
                      todo.status === TodoStatus.COMPLETED
                        ? Styles.completed
                        : Styles.active
                    }
                  >
                    {todo.content}
                  </span>
                )}
              </div>
              <Button
                className={Styles.remove}
                type='button'
                onClick={() => onDeleteTodo(todo.id)}
              >
                X
              </Button>
            </div>
          );
        })
      ) : (
        <div className={Styles.recondFound}>No record found</div>
      )}
    </div>
  );
};

export default TodoList;
