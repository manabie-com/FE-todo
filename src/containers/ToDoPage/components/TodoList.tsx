import React from 'react';
import styled from 'styled-components';

import { TodoItem } from './TodoItem';
import { EntityId } from '@reduxjs/toolkit';

interface Props {
  ids: EntityId[];
}

export function TodoList(props: Props) {
  const { ids } = props;

  const count = ids.length;
  return (
    <Wrapper>
      {count <= 0 ? (
        <div>Nothing to show</div>
      ) : (
        ids.map(id => <TodoItem key={id} id={id} />)
      )}
    </Wrapper>
  );
}

const Wrapper = styled.div`
  padding: 24px 0;
`;
