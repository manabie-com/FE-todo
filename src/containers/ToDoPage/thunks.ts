import { createAsyncThunk } from '@reduxjs/toolkit';
import Service from 'service';

export const createTodo = createAsyncThunk(
  'todoPage/createTodo',
  async (content: string) => {
    const todo = await Service.createTodo(content);
    return todo;
  },
);
