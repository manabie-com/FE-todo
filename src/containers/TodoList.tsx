import React, { useContext } from 'react'
import { TodoContext } from '../contexts/todo'
import { deleteTodo, updateTodo } from '../store/actions'
import { getTodoItemsByStatus, isTodoCompleted } from '../utils'
import InlineEditable from '../components/InlineEditable'
import { Todo, TodoStatus } from '../models/todo'
import Service from '../service'

interface TodoListProps {
    status: string
}

const TodoList = (props: TodoListProps) => {
    const { store, dispatch } = useContext(TodoContext)
    const renderItems = getTodoItemsByStatus(props.status, store.todos)

    const onItemCheck = (todo: Todo) => async (e: React.ChangeEvent<HTMLInputElement>) => {
        const newTodo = {
            ...todo,
            status: e.target.checked ? TodoStatus.COMPLETED : TodoStatus.ACTIVE
        } as Todo
        await Service.updateTodoStatus([newTodo])
        dispatch(updateTodo(newTodo))
    }

    const onEditItemName = (todo: Todo) => async (content: string) => {
        const newItem = { ...todo, content }
        await Service.updateTodoItemContent(newItem)
        dispatch(updateTodo(newItem))
    }

    const onDeleteTodoItem = (todo: Todo) => async () => {
        await Service.deleteTodoItem(todo.id)
        dispatch(deleteTodo(todo.id))
    }

    return (
        <div className="ToDo__list">
            {
                renderItems.map(
                    todoItem =>
                        <div key={todoItem.id} className="ToDo__item">
                            <input
                                type="checkbox"
                                className="Todo__cbx"
                                checked={isTodoCompleted(todoItem)}
                                onChange={onItemCheck(todoItem)}
                            />
                            <InlineEditable todoName={todoItem.content} onChange={onEditItemName(todoItem)}/>
                            <button
                                className="Todo__delete"
                                onClick={onDeleteTodoItem(todoItem)}
                            >
                                X
                            </button>
                        </div>
                )
            }
        </div>
    )
}

export default TodoList
