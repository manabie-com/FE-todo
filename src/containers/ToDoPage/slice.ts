import {
  createSlice,
  createEntityAdapter,
  PayloadAction,
} from '@reduxjs/toolkit';

import { Todo, TodoStatus } from 'models/todo';
import { RootState } from 'store/RootState';
import { createTodo } from './thunks';

export const adapter = createEntityAdapter<Todo>();

export const initialState = adapter.getInitialState({
  filter: 'ALL',
});

const todoPageSlice = createSlice({
  name: 'todoPage',
  initialState,
  reducers: {
    deleteTodo: adapter.removeOne,
    deleteAllTodos: adapter.removeAll,
    updateTodo: adapter.updateOne,
    completeAllTodos(state) {
      state.ids.forEach(id => {
        const todo = state.entities[id];
        if (todo) {
          todo.status = TodoStatus.COMPLETED;
        }
      });
    },
    setFilter(state, action: PayloadAction<string>) {
      state.filter = action.payload;
    },
  },
  extraReducers: builder => {
    builder.addCase(createTodo.fulfilled, adapter.addOne);
  },
});

export const { actions, reducer, name: sliceKey } = todoPageSlice;
export const {
  selectById: selectTodoById,
  selectIds: selectTodoIds,
} = adapter.getSelectors<RootState>(state => state.todoPage);
