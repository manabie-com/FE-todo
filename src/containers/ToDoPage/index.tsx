import React from 'react';
import { Redirect } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components/macro';

import { Button } from 'components/Button';
import { Form } from './components/Form';
import { TodoList } from './components/TodoList';
import { actions, selectTodoIds } from './slice';
import { isLoggedIn } from 'utils/auth';

export function ToDoPage() {
  const todoIds = useSelector(selectTodoIds);
  const dispatch = useDispatch();

  const handleFilter = (filter: string) => () => {
    dispatch(actions.setFilter(filter));
  };

  const handleCompleteAll = () => {
    dispatch(actions.completeAllTodos());
  };

  const handleRemoveAll = () => {
    dispatch(actions.deleteAllTodos());
  };

  if (!isLoggedIn()) {
    return <Redirect to="/" />;
  }

  return (
    <Wrapper>
      <Form />

      <TodoList ids={todoIds} />

      <Actions>
        <Button onClick={handleFilter('ALL')}>Show all</Button>
        <Button onClick={handleFilter('ACTIVE')}>Show active</Button>
        <Button onClick={handleFilter('COMPLETED')}>Show completed</Button>
        <Button primary onClick={handleCompleteAll}>
          Complete all
        </Button>
        <Button primary onClick={handleRemoveAll}>
          Delete all
        </Button>
      </Actions>
    </Wrapper>
  );
}

const Wrapper = styled.div`
  box-shadow: 0 1px 2px 0 rgba(34, 36, 38, 0.15);
  border: 1px solid rgba(34, 36, 38, 0.15);
  margin: 200px auto;
  max-width: 600px;
  padding: 16px;
  border-radius: 6px;
`;

const Actions = styled.div`
  display: flex;
  justify-content: space-between;
`;
