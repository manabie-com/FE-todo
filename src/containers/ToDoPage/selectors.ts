import { createSelector } from '@reduxjs/toolkit';
import { RootState } from 'store/RootState';
import { initialState, selectTodoById } from './slice';

const selectDomain = (state: RootState) => state.todoPage || initialState;

export const selectFilter = createSelector(
  [selectDomain],
  todoPageState => todoPageState.filter,
);

export const selectVisibleTodo = createSelector(
  [selectTodoById, selectFilter],
  (todo, filter) => {
    if (!todo) return null;

    if (filter !== 'ALL' && todo.status !== filter) return null;

    return todo;
  },
);
