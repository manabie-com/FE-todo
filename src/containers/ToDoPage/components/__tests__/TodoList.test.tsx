import * as React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import { configureAppStore } from 'store/configureStore';
import { createTodo } from 'containers/ToDoPage/thunks';
import { TodoList } from '../TodoList';

describe('<TodoList />', () => {
  it('should show empty message', () => {
    const { getByText } = render(<TodoList ids={[]} />);
    expect(getByText('Nothing to show')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const store = configureAppStore();
    const todo = { id: 1, content: 'test' };
    store.dispatch({ type: createTodo.fulfilled.type, payload: todo });
    const { container } = render(
      <Provider store={store}>
        <TodoList ids={[1]} />
      </Provider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
