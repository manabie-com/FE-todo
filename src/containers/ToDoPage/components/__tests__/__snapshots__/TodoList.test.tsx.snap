// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> should match snapshot 1`] = `
.c8 {
  outline: none;
  cursor: pointer;
  border: none;
  font-weight: 700;
  padding: 8px 16px;
  border-radius: 6px;
  background-color: #e0e1e2;
  color: rgba(0,0,0,0.6);
}

.c8:hover,
.c8:focus {
  background-color: #cacbcd;
  color: rgba(0,0,0,0.8);
}

.c8:active {
  background-color: #babbbc;
  color: rgba(0,0,0,0.9);
}

.c2 {
  position: relative;
  min-width: 24px;
  min-height: 24px;
}

.c2::before {
  content: '';
  position: absolute;
  z-index: -1;
  box-sizing: border-box;
  width: 24px;
  height: 24px;
  border-radius: 4px;
  border: 1px solid #2185d0;
}

.c5 {
  width: 24px;
  color: #2185d0;
  position: absolute;
  z-index: -1;
}

.c3 {
  position: absolute;
  opacity: 0;
  width: 24px;
  height: 24px;
  margin: 0;
}

.c3 ~ .c4 {
  display: none;
}

.c3:checked ~ .c4 {
  display: block;
}

.c6 {
  padding-left: 36px;
  line-height: 24px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:not(:last-child) {
  margin-bottom: 8px;
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c0 {
  padding: 24px 0;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <input
        class="c3"
        id="1"
        type="checkbox"
      />
      <svg
        class="c4 c5"
      >
        check.svg
      </svg>
      <label
        class="c6"
        for="1"
      />
    </div>
    <div
      class="c7"
    >
      test
    </div>
    <button
      class="c8"
    >
      x
    </button>
  </div>
</div>
`;
